File: Baekjoon/단계별/20단계 - 스택/4949번 - 균형잡힌 세상.cpp
#include <iostream>
#include <algorithm>
#include <string>
#include <stack>

using namespace std;

string s;

string checkVPS(string input) {
	stack<char> st;

	for (int i = 0; i < input.length(); i++) {
		if (input[i] == '(' || input[i] == '[') {
			st.push(input[i]);
		}
		else if (input[i] == ')') {
			if (!st.empty() && st.top() == '(')
				st.pop();
			else
			{
				return "no";
			}
		}
		else if (input[i] == ']') {
			if (!st.empty() && st.top() == '[')
				st.pop();
			else
			{
				return "no";
			}
		}
	}

	if (st.empty())
		return "yes";
	else
		return "no";
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	while (1) {
		getline(cin, s);

		if (s[0] == '.')
			break;

		cout << checkVPS(s) << "\n";
	}

	return 0;
}	

File: 백준/Python/Silver/4949번 - 균형잡힌 세상.py
import sys

while True:
    str = sys.stdin.readline()
    if str == ".":
        break

    stack = []

    for i in str:
        if i == '(' or i == '[':
            stack.append(i)
        elif i == ')': 
            if len(stack) != 0 and stack[-1] == '(':
                stack.pop()
            else:
                stack.append(i)
                break 
        elif i == ']':
            if len(stack) != 0 and stack[-1] == '[':
                stack.pop()
            else:
                stack.append(i)
                break 

    if not stack:
        print('yes')
    else:
        print('no')